# step 1

setup the data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot
from sklearn.model_selection import train_test_split

df=pd.read_csv(r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\week3\day3\WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.isna().sum()
df.columns
df.dtypes
df['Dependents']= df['Dependents'].map({
    'Yes':1,
    'No':0
})
df['Partner']= df['Partner'].map({
    'Yes':1,
    'No':0
})
df=df[['SeniorCitizen','Partner','Dependents','tenure','MonthlyCharges','Contract','Churn']]
y=df['Churn']
x=df.drop('Churn',axis=1)
x=pd.get_dummies(x)
x.columns
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)


### Decision Trees

In [2]:
from sklearn.tree import DecisionTreeClassifier
model= DecisionTreeClassifier(max_depth=5,random_state=42)
model.fit(X_train,y_train)
pred=model.predict(X_test)

### Random Forests

In [3]:
from sklearn.ensemble import RandomForestClassifier
model1=RandomForestClassifier(n_estimators=100,random_state=42)
model1.fit(X_train,y_train)
pred1=model1.predict(X_test)

### Support Vector Machines (SVM)


In [4]:
from sklearn.svm import SVC
model2=SVC(kernel='rbf',probability=True)
model2.fit(X_train,y_train)
pred2=model2.predict(X_test)

c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(


In [5]:
from sklearn.neighbors import KNeighborsClassifier
model3=KNeighborsClassifier(n_neighbors=5)
model3.fit(X_train,y_train)
pred3=model3.predict(X_test)

# step 2

compare by f1

In [7]:
from sklearn.metrics import f1_score

results = pd.DataFrame({
    "Model": [
        "Decision Tree",
        "Random Forest",
        "SVM",
        "k-NN"
    ],
    "F1-score": [
        f1_score(y_test, pred, pos_label="Yes"),
        f1_score(y_test, pred1, pos_label="Yes"),
        f1_score(y_test, pred2, pos_label="Yes"),
        f1_score(y_test, pred3, pos_label="Yes")
    ]
})

results

,Model,F1-score
0,Decision Tree,0.579915
1,Random Forest,0.531025
2,SVM,0.522920
3,k-NN,0.522388


# step 3

In [13]:
feature_importance = pd.DataFrame({
    "Feature": X_train.columns,
    "Importance": model1.feature_importances_
})
feature_importance

,Feature,Importance
0,SeniorCitizen,0.015212
1,Partner,0.012950
2,Dependents,0.014595
3,tenure,0.284953
4,MonthlyCharges,0.539074
5,Contract_Month-to-month,0.086456
6,Contract_One year,0.015170
7,Contract_Two year,0.031590


### Interpretation of Feature Importances:

"The Random Forest model identified MonthlyCharges (53.9%) and tenure (28.5%) as the two most critical drivers for prediction, accounting for over 82% of the model's decision weight. Additionally, contract type (specifically Month-to-month) played a notable role, while demographic attributes (Partner, Dependents, SeniorCitizen) had minimal impact on the prediction outcome

# step 4

### Model Comparison and Winner Analysis:

"After evaluating all four models on the same test split using the F1-score metric, the Decision Tree model achieved the highest performance with an F1-score of 0.580, outperforming the Random Forest (0.531), SVM (0.523), and k-NN (0.522).

Why it likely won:
Although Ensemble models like Random Forests are generally robust, this specific dataset appears to be governed by clear, direct, and straightforward logical rules that a single Decision Tree can capture efficiently without severe overfitting. On the other hand, distance-based models (SVM and k-NN) struggled slightly more with the dataset's specific feature space, proving the 'no free lunch' principle: the simplest model can sometimes outperform more complex ones depending on the underlying data distribution."